htmd/qm/base.py

```diff
@@ -0,0 +1,202 @@
+        self.quadrupole = None
+        self.mulliken = None
+        self.esp_points = None
+        self.esp_values = None
+
+
+class QMBase(ABC, ProtocolInterface):
+    """
+    Abstract base class to set up and run QM calculations
+    """
+
+    THEORIES = ('HF', 'BLYP', 'PBE', 'B3LYP', 'PBE0', 'B2PLYP')
+    CORRECTIONS = ('none', 'D', 'D3')
+    BASIS_SETS = ('3-21G',
+                  '6-31G',  '6-31G*',  '6-31G**',  '6-31+G',  '6-31+G*',  '6-31+G**',  '6-31++G',  '6-31++G*',  '6-31++G**',
+                  '6-311G', '6-311G*', '6-311G**', '6-311+G', '6-311+G*', '6-311+G**', '6-311++G', '6-311++G*', '6-311++G**',
+                  'cc-pVDZ', 'cc-pVTZ', 'cc-pVQZ', 'aug-cc-pVDZ', 'aug-cc-pVTZ', 'aug-cc-pVQZ')
+    SOLVENTS = ('vacuum', 'PCM')
+
+    def __init__(self):
+
+        from htmd.parameterization.ffmolecule import FFMolecule
+
+        super().__init__()
+
+        self._arg('molecule', ':class: `htmd.parameterization.ffmolecule.FFMolecule`', 'Molecule',
+                  validator=val.Object(FFMolecule), required=True)
+        self._arg('multiplicity', 'int', 'Multiplicity of the molecule',
+                  default=1, validator=val.Number(int, 'POS'))
+        self._arg('theory', 'str', 'Level of theory',
+                  default='B3LYP', validator=val.String(), valid_values=self.THEORIES)
+        self._arg('correction', 'str', 'Empirical dispersion correction',
+                  default='none', validator=val.String(), valid_values=self.CORRECTIONS)
+        self._arg('basis', 'str', 'Basis set',
+                  default='6-31G*', validator=val.String(), valid_values=self.BASIS_SETS)
+        self._arg('solvent', 'str', 'Implicit solvent',
+                  default='vacuum', validator=val.String(), valid_values=self.SOLVENTS)
+        self._arg('esp_points', ':class: `numpy.ndarray`', 'Point to calculate ESP',
+                  default=None, nargs='*')  # TODO implement validator
+        self._arg('optimize', 'boolean', 'Optimize geometry',
+                  default=False, validator=val.Boolean())
+        self._arg('restrained_dihedrals', ':class: `numpy.ndarray`',
+                  'List of restrained dihedrals (0-based indices)',
+                  default=None, nargs='*')  # TODO implement validator
+        self._arg('queue', ':class:`SimQueue <htmd.queues.simqueue.SimQueue>` object',
+                  'Queue object used to run simulations',
+                  default=LocalCPUQueue())
+        self._arg('directory', 'str', 'Working directory',
+                  default='.', validator=val.String())
+
+    @property
+    @abstractmethod
+    def _command(self):
+        pass
+
+    @abstractmethod
+    def _completed(self, directory):
+        pass
+
+    @abstractmethod
+    def _writeInput(self, directory, iframe):
+        pass
+
+    @abstractmethod
+    def _readOutput(self, directory):
+        pass
+
+    def _setup(self):
+
+        # Set up the molecule
+        # TODO remove molecule coping!
```

Comment:
Why the copy?

Reply:
Because QM needed to reorient the molecule. It is not case any more for Psi4, but I'm not sure if Gaussian can avoid doing that.